### PR TITLE
Add package relocation.

### PR DESCRIPTION
**Pull request by mamokarz**

### CODE REVIEW

File: STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/az_ulib_dm.c [hunk 1]
```diff
@@ -97,16 +98,35 @@ static az_result install_in_memory(void* base_address, az_span package_name)
     _az_ulib_dm_package* package = get_first_package_free();
     AZ_ULIB_THROW_IF_ERROR((package != NULL), AZ_ERROR_NOT_ENOUGH_SPACE);
 
-    /* Call the publsh interface function in the package. */
-    uint32_t publish_position = *((uint32_t*)base_address + AZ_ULIB_DM_PACKAGE_PUBLISH);
+    /* Check it the package name fits in the reserved buffer. */
```
Comment (smcd253) [inline STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/az_ulib_dm.c]: >it [](http://example.com/codeflow?start=13&length=2)

"if" #Closed

File: STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/az_ulib_dm_blob_ustream_interface.c [hunk 2]
```diff
@@ -106,93 +106,95 @@ static az_result concrete_read(
...
-      result = _az_nx_blob_client_grab_chunk(
-          &blob_http_cb->_internal.http_client,
-          &blob_http_cb->_internal.packet_ptr,
-          BLOB_CLIENT_NX_API_WAIT_TIME);
-      if ((result == AZ_OK) || (result == AZ_ULIB_EOF))
+        // set result to AZ_OK for next loop
+        result = AZ_OK;
+      }
+
+      // All bytes in the current packet was consumed.
```
Comment (smcd253) [inline STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/az_ulib_dm_blob_ustream_interface.c]: >was [](http://example.com/codeflow?start=41&length=3)

"were" #Closed

File: STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/az_ulib_dm_blob_ustream_interface.c [hunk 3]
```diff
@@ -284,26 +286,20 @@ AZ_NODISCARD az_result az_blob_create_ustream_from_blob(
   _az_PRECONDITION_NOT_NULL(ustream_data_cb);
   _az_PRECONDITION_NOT_NULL(blob_http_cb);
   _az_PRECONDITION_NOT_NULL(ip);
-  _az_PRECONDITION(strlen((char*)resource) > 0);
-  _az_PRECONDITION(strlen((char*)host) > 0);
+  _az_PRECONDITION_NOT_NULL(resource);
+  _az_PRECONDITION_NOT_NULL(host);
+  _az_PRECONDITION(resource[0] != 0);
```
Comment (smcd253) [inline STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/az_ulib_dm_blob_ustream_interface.c:291]: >_az_PRECONDITION(resource[0] != 0) [](http://example.com/codeflow?start=2&length=34)

what's the motivation behind checking the first index like this? #ByDesign

File: STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/az_ulib_dm_blob_ustream_interface.c [hunk 4]
```diff
@@ -314,10 +310,6 @@ AZ_NODISCARD az_result az_blob_create_ustream_from_blob(
     // grab blob package size for ustream init
     uint32_t package_size
         = (uint32_t)blob_http_cb->_internal.http_client.nx_web_http_client_total_receive_bytes;
-    if ((package_size & 0x07FF) != 0x000)
-    {
-      package_size = (package_size & 0xFFFFF800) + 0x0800;
-    }
```
Comment (smcd253) [inline STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/az_ulib_dm_blob_ustream_interface.c:320]: Is there no need to round to the nearest 2K anymore? #ByDesign

File: STMicroelectronics/STM32L4_L4+/samples/key_vault.1/key_vault_interface.c [hunk 5]
```diff
@@ -3,7 +3,7 @@
 // See LICENSE file in the project root for full license information.
 
 /********************************************************************
- * This code was auto-generated from key_vault.1 DL and shall not be
+ * This code was auto-generated from cipher.1 DL and shall not be
```
Comment (smcd253) [inline STMicroelectronics/STM32L4_L4+/samples/key_vault.1/key_vault_interface.c]: >cipher.1 DL [](http://example.com/codeflow?start=37&length=11)

I think this has potential to be a tad confusing. cipher is a capability within the key_vault package but the name of this source file is only generated from the cipher model.  #Closed

File: STMicroelectronics/STM32L4_L4+/samples/key_vault.1/startup/gnu/gcc_setup.s [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+    add     r6, r6, r5      // Build address based on the loaded code address
+address_built:
+    str     r6, [r1]        // Store in new GOT table
+    add     r0, r0, #4      // Move to next entry
+    add     r1, r1, #4      // 
+    b       new_got_setup   // Continue at the top of the loop
+got_setup_done:
+
+
+    /* Copy initialised sections into RAM if required. */
```
Comment (smcd253) [inline STMicroelectronics/STM32L4_L4+/samples/key_vault.1/startup/gnu/gcc_setup.s]: >initialised [](http://example.com/codeflow?start=12&length=11)

"initialized" #Closed

File: STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/az_ulib_dm_blob_ustream_interface.c (hunk 3 above)
Comment (mamokarz) [reply STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/az_ulib_dm_blob_ustream_interface.c:291]: I'm replacing the strlen by 2 tests: first, test if the pointer is not NULL, and second, test if there is, at least, one non-null byte in the string. It is a better solution because if the pointer is NULL, strlen will crash instead of return 0, and strlen will run over the entire string but if the first char is non-null, the string is not empty.

File: STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/az_ulib_dm_blob_ustream_interface.c (hunk 4 above)
Comment (mamokarz) [reply STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/az_ulib_dm_blob_ustream_interface.c:320]: 1. This is particular of the consumer of this stream. You cannot change the size of a stream to fit the needs of the consumer of it, if the consumer needs to resize it, the consumer shall do that by its own. 

2. The Flash APIs requires the real number of bytes that will be written. The eraser API will handle the block size, so you don't need to do that.

3. If you look at the API write in Flash, it relies on the right size of the file to write the last bytes, in the case that the size is not aligned to 8 bytes, change the size here was causing a problem with the last bytes.

File: STMicroelectronics/STM32L4_L4+/samples/key_vault.1/key_vault_interface.c (hunk 5 above)
Comment (mamokarz) [reply STMicroelectronics/STM32L4_L4+/samples/key_vault.1/key_vault_interface.c]: You are right, I need to change the name of this file and remove all reference to key_vault from it. I'll rollback this change and fix it in another PR.

File: STMicroelectronics/STM32L4_L4+/create_package.md [hunk 7]
```diff
@@ -4,52 +4,42 @@ The following steps create a new package to be installed in the device. This pac
...
 ```c
 // Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license.
 // See LICENSE file in the project root for full license information.
 
+/********************************************************************
+ * This code was auto-generated from sprinkler.1 DL and shall not be
+ * modified.
+ ********************************************************************/
+
```
Comment (mamokarz) [inline STMicroelectronics/STM32L4_L4+/create_package.md]: Remove #Closed

File: STMicroelectronics/STM32L4_L4+/create_package.md [hunk 8]
```diff
@@ -100,24 +90,29 @@ extern "C"
...
```
Comment (mamokarz) [inline STMicroelectronics/STM32L4_L4+/create_package.md]: Remove #Closed

File: STMicroelectronics/STM32L4_L4+/create_package.md [hunk 9]
```diff
@@ -430,53 +458,31 @@ At this point, your sprinkler_v2i1 project tree shall looks like:
                 tx_initialize_low_level.S
 ```
 
-This CMake will create an executable sprinkler_v2i1.elf that contains the files in the startup directory plus the concrete implementation of the interface and the business logic. Because it will expose an IPC interface, the library "az::ulib" is required. The library "az::core" brings the implementation for the JSON parser, and "stm32cubel4" the HAL to access the MCU pins. 
+This CMake will create an executable sprinkler.2.elf that contains the files in the startup directory plus the concrete implementation of the interface and the business logic. Because it will expose an IPC interface, the library "az::ulib" is required. The library "az::core_pic" brings the implementation for the JSON parser, and "stm32cubel4" the HAL to access the MCU pins. 
+
+Sprinkler.1.elf contains the binary code to run in the MCU together with a few more information, tools like jTag knows that and extract the binary when loading the .elf to the MCU. In our case, we need to extract the binary to write directly in the memory. The POST_BUILD ObjCopy do this extraction creating the sprinkler.2.bin.
```
Comment (mamokarz) [inline STMicroelectronics/STM32L4_L4+/create_package.md]: >1 [](http://example.com/codeflow?start=10&length=1)

2 #Closed

File: STMicroelectronics/STM32L4_L4+/create_package.md [hunk 10]
```diff
@@ -430,53 +458,31 @@ At this point, your sprinkler_v2i1 project tree shall looks like:
...
 
-Sprinkler_v1i1.elf contains the binary code to run in the MCU together with a few more information, tools like jTag knows that and extract the binary when loading the .elf to the MCU. In our case, we need to extract the binary to write directly in the memory. The POST_BUILD ObjCopy do this extraction creating the sprinkler_v2i1.bin.
+Include the new sample in the CMake of the samples by adding `add_subdirectory(sprinkler.2)` in the samples/CMakeLists.txt.
 
-To build the project, call again ".\tools\rebuild.bat". It will build the entire project creating 4 executables, including the new sprinkler_v2i1.elf
+To build the project, call again ".\tools\rebuild.bat". It will build the entire project creating 6 executables, including the new sprinkler.2.elf
```
Comment (mamokarz) [inline STMicroelectronics/STM32L4_L4+/create_package.md]: >6 [](http://example.com/codeflow?start=98&length=1)

all #Closed

File: STMicroelectronics/STM32L4_L4+/create_package.md [hunk 11]
```diff
@@ -430,53 +458,31 @@ At this point, your sprinkler_v2i1 project tree shall looks like:
...
-
-Use the same steps as earlier to flash this new package into address `0x08057000` with the ST-Link Utility tool. Reference this link [How to flash new package into device memory with ST-Link](../../DCF_Demo.md###STM32-ST-Link-Utility) from earlier. 
+Use the same steps as earlier to flash this new package (for example, into address `0x08057000`) with the [GDB commands](../../DCF_Demo.md###GDB-Commands) or [ST-Link Utility tool](../../DCF_Demo.md###STM32-ST-Link-Utility).
 
 ## Test the New Package
-Now, similar to key_vault_v1i1, you can install sprinkler_v2i1 using `invoke-device-method`. To install sprinker_v2i1 package in the address 134574080 [0x08057000] 
+Now, similar to key_vault.1, you can install sprinkler.2 using `invoke-device-method`. To install sprinker.2 package in the address 134574080 [0x08057000] 
 ```
-az iot hub invoke-device-method -n [name-of-iothub] -d [name-of-device] --mn "dm.1.install" --mp "{\"source_type\":0,\"address\":134574080,\"package_name\":\"sprinker_v2i1\"}" 
+az iot hub invoke-device-method -n [name-of-iothub] -d [name-of-device] --mn "dm.1.install" --mp "{\"source_type\":0,\"address\":134574080,\"package_name\":\"sprinker.2\"}"
```
Comment (mamokarz) [inline STMicroelectronics/STM32L4_L4+/create_package.md]: >dm.1 [](http://example.com/codeflow?start=78&length=4)

dm.packages.1 #Closed

File: DCF_Demo.md [hunk 12]
```diff
@@ -6,26 +6,32 @@ Please make sure you have completed the Getting Started Guide for the STM32 boar
 ## Open up an instance of Azure CLI 
 You can use the Azure Portal or an instance on Powershell/bash/other local environments
 
-Device Manager current support install code from 4 different sources.
+Device Manager current support install code from 2 different sources.
```
Comment (smcd253) [inline DCF_Demo.md]: >current support [](http://example.com/codeflow?start=15&length=15)

"currently supports installing" #Closed

File: DCF_Demo.md [hunk 13]
```diff
@@ -6,26 +6,32 @@ Please make sure you have completed the Getting Started Guide for the STM32 boar
...
 0. **In memory**: Is the code that was brought to the memory using external interfaces (like JTag).
 1. **From Blobs**: DM will download the code from Blobs using the provided URL.
-2. **From CLI** (*not supported yet*): DM will copy the code from CLI to the memory.
 
-## Commands to Run 
-
-### In Memory example
+## In Memory example
 
 Install an **In Memory** package assume that the binary is already in the MCU memory when the install command arrives. The main goal of this installation is to debug package's code.
```
Comment (smcd253) [inline DCF_Demo.md]: >Install an **In Memory** package assume  [](http://example.com/codeflow?start=0&length=40)

nit: "Installing an **In Memory** package assumes" #Closed

File: DCF_Demo.md [hunk 14]
```diff
@@ -6,26 +6,32 @@ Please make sure you have completed the Getting Started Guide for the STM32 boar
...
-## Commands to Run 
-
-### In Memory example
+## In Memory example
 
 Install an **In Memory** package assume that the binary is already in the MCU memory when the install command arrives. The main goal of this installation is to debug package's code. 
 If properly configured, all debug tools applicable to a code built in the OS can be equally applied to a package using the DCF in memory installation.
 
-The current example produces the key_vault_v1i1 and sprinkler_v1i1 packages. The CMake will create the key_vault_v1i1.elf and key_vault_v1i1.bin for key_vault_v1i1 package in ".\build\samples\key_vault_v1i1" directory, and sprinkler_v1i1.elf and sprinkler_v1i1.bin for sprinkler_v1i1 package in ".\build\samples\sprinkler_v1i1" directory. 
+The current example produces the key_vault.1, key_vault.2 and sprinkler.1 packages. The CMake will create the `.elf` and `.bin` for each package in ".\build\samples" directory.
```
Comment (smcd253) [inline DCF_Demo.md]: >in  [](http://example.com/codeflow?start=145&length=3)

nit "in the" #Closed

File: DCF_Demo.md [hunk 15]
```diff
@@ -6,26 +6,32 @@ Please make sure you have completed the Getting Started Guide for the STM32 boar
...
+  .\build\samples\key_vault.1\key_vault.1.elf
+  .\build\samples\key_vault.1\key_vault.1.bin
+  .\build\samples\key_vault.2\key_vault.2.elf
+  .\build\samples\key_vault.2\key_vault.2.bin
+  .\build\samples\sprinkler.1\sprinkler.1.elf
+  .\build\samples\sprinkler.1\sprinkler.1.bin
+```
 
-## Upload Binary to MCU Flash
+All packages use Position Independent Code (PIC), so they can be installed in any available memory in the MCU, for example, you may install key_vault.1 in the address 0x08050000, and sprinkler.1 in the address 0x08057000.
```
Comment (smcd253) [inline DCF_Demo.md]: >MCU, for example [](http://example.com/codeflow?start=106&length=16)

nit: "... MCU. For example," #Closed

File: DCF_Demo.md [hunk 16]
```diff
@@ -6,26 +6,32 @@ Please make sure you have completed the Getting Started Guide for the STM32 boar
...
 
-Here are two ways to upload the binary to the MCU flash. Both examples use key_vault_v1i1, you can do the same for the sprinkler_v1i1 replacing the package name and the target address from 0x08050000 by 0x08057000.  
+### Upload Binary to MCU Flash
+
+Here are two ways to upload the binary to the MCU flash. Both examples use key_vault.1, you can do the same for key_vault.2 and sprinkler.1, replacing the package name and the target address.
```
Comment (smcd253) [inline DCF_Demo.md]: >key_vault.1, you [](http://example.com/codeflow?start=75&length=16)

nit: "... key_vault.1. You ..." #Closed

File: DCF_Demo.md [hunk 17]
```diff
@@ -34,11 +40,49 @@ Here are two ways to upload the binary to the MCU flash. Both examples use key_v
...
 
 ```
-restore build/samples/key_vault_v1i1/key_vault_v1i1.bin binary 0x08050000
-```
-and, if you are using GDB to debug your code, you can add the symbols as well. 
-```
-add-symbol-file build/samples/key_vault_v1i1/key_vault_v1i1.elf 0x08050080
+load build/samples/key_vault.1/key_vault.1.elf 0x08050000
+```
+and, if you are using GDB to debug your code, you can add the symbols as well. Symbols shall be attached from the .text section. to find it, you can use the readelf tool.
```
Comment (smcd253) [inline DCF_Demo.md]: >and, if [](http://example.com/codeflow?start=0&length=7)

nit: remove "and,". "If you are using..." #Closed

File: DCF_Demo.md [hunk 18]
```diff
@@ -34,11 +40,49 @@ Here are two ways to upload the binary to the MCU flash. Both examples use key_v
 One of the ways to upload the binary file is to use GDB commands.
```
Comment (smcd253) [inline DCF_Demo.md]: >use GDB commands. [](http://example.com/codeflow?start=48&length=17)

nit: "use the following GDB command:" #Closed

File: DCF_Demo.md [hunk 19]
```diff
@@ -34,11 +40,49 @@ Here are two ways to upload the binary to the MCU flash. Both examples use key_v
...
+  [22] .debug_aranges    PROGBITS        00000000 045860 000448 00      0   0  8
+  [23] .debug_str        PROGBITS        00000000 045ca8 00dac7 01  MS  0   0  1
+  [24] .debug_ranges     PROGBITS        00000000 05376f 000418 00      0   0  1
+  [25] .debug_macro      PROGBITS        00000000 053b87 003797 00      0   0  1
+  [26] .debug_frame      PROGBITS        00000000 057320 001230 00      0   0  4
+  [27] .symtab           SYMTAB          00000000 058550 001500 10     28 243  4
+  [28] .strtab           STRTAB          00000000 059a50 000f14 00      0   0  1
+  [29] .shstrtab         STRTAB          00000000 05a964 000112 00      0   0  1
+```
+As a result, you can see in the section 9, that .text starts at the relative position 0x00000260, which we shall add to the installed address 0x08050000, resulting in the address 0x08050260. The address may be different in your compilation.
```
Comment (smcd253) [inline DCF_Demo.md]: >in the section 9, that .text [](http://example.com/codeflow?start=25&length=28)

nit: "in Section 9 that `.text`" #Closed

File: DCF_Demo.md [hunk 20]
```diff
@@ -309,14 +357,18 @@ az iot hub invoke-device-method -n [name-of-iothub] -d [name-of-device] --mn "ke
 </details>
 <br>
 
-### From Blob example
+## From Blob example
 
 Once you have your package ready, it is time to distribute it to the devices. 
 
 If you do not have a Azure Storage account, you may create one following the steps in the [Create a storage account](https://docs.microsoft.com/en-us/azure/storage/common/storage-account-create?toc=%2Fazure%2Fstorage%2Fblobs%2Ftoc.json&tabs=azure-portal).
```
Comment (smcd253) [inline DCF_Demo.md]: >a [](http://example.com/codeflow?start=19&length=1)

nit: "an" #Closed

File: DCF_Demo.md [hunk 21]
```diff
@@ -309,14 +357,18 @@ az iot hub invoke-device-method -n [name-of-iothub] -d [name-of-device] --mn "ke
 </details>
 <br>
 
-### From Blob example
+## From Blob example
```
Comment (smcd253) [inline DCF_Demo.md:360]: >From Blob example [](http://example.com/codeflow?start=3&length=17)

do you plan on covering default packages in this demo? #WontFix

File: STMicroelectronics/STM32L4_L4+/samples/key_vault.1/CMakeLists.txt [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+)
+
+add_custom_command(TARGET key_vault.1 
+    POST_BUILD
+        COMMAND ${CMAKE_OBJCOPY} ARGS -O binary -j .preamble -j .dynsym -j dynstr -j .rel.dyn 
+          -j .rel.plt -j .plt -j .interp -j .hash -j .text -j .dtors -j .ctors -j .got -j .rodata 
+          -j .fast -j .fast_run -j .data -j .data_run -j .bss -j .non_init -j .heap
+        ${CMAKE_CURRENT_LIST_DIR}/../../build/samples/key_vault.1/key_vault.1.elf
+        ${CMAKE_CURRENT_LIST_DIR}/../../build/samples/key_vault.1/key_vault.1.bin
+    COMMENT "Converting the ELF output to a binary file"
```
Comment (smcd253) [inline STMicroelectronics/STM32L4_L4+/samples/key_vault.1/CMakeLists.txt]: could be useful to have a source in comments here so developers can see how this was done #Closed

File: STMicroelectronics/STM32L4_L4+/samples/key_vault.2/CMakeLists.txt [hunk 23]
```diff
@@ -0,0 +1,50 @@
...
+
+add_custom_command(TARGET key_vault.2 
+    POST_BUILD
+        COMMAND ${CMAKE_OBJCOPY} ARGS -O binary -j .preamble -j .dynsym -j dynstr -j .rel.dyn 
+          -j .rel.plt -j .plt -j .interp -j .hash -j .text -j .dtors -j .ctors -j .got -j .rodata 
+          -j .fast -j .fast_run -j .data -j .data_run -j .bss -j .non_init -j .heap
+        ${CMAKE_CURRENT_LIST_DIR}/../../build/samples/key_vault.2/key_vault.2.elf
+        ${CMAKE_CURRENT_LIST_DIR}/../../build/samples/key_vault.2/key_vault.2.bin
+    COMMENT "Converting the ELF output to a binary file"
+)
```
Comment (smcd253) [inline STMicroelectronics/STM32L4_L4+/samples/key_vault.2/CMakeLists.txt:50]: same as key_vault.1 CMakeLists.txt comment #Closed

File: STMicroelectronics/STM32L4_L4+/samples/sprinkler.1/CMakeLists.txt [hunk 24]
```diff
@@ -0,0 +1,51 @@
...
+
+add_custom_command(TARGET sprinkler.1 
+    POST_BUILD
+        COMMAND ${CMAKE_OBJCOPY} ARGS -O binary -j .preamble -j .dynsym -j dynstr -j .rel.dyn 
+          -j .rel.plt -j .plt -j .interp -j .hash -j .text -j .dtors -j .ctors -j .got -j .rodata 
+          -j .fast -j .fast_run -j .data -j .data_run -j .bss -j .non_init -j .heap
+        ${CMAKE_CURRENT_LIST_DIR}/../../build/samples/sprinkler.1/sprinkler.1.elf
+        ${CMAKE_CURRENT_LIST_DIR}/../../build/samples/sprinkler.1/sprinkler.1.bin
+    COMMENT "Converting the ELF output to a binary file"
+)
```
Comment (smcd253) [inline STMicroelectronics/STM32L4_L4+/samples/sprinkler.1/CMakeLists.txt:51]: same here #Closed

File: STMicroelectronics/STM32L4_L4+/samples/sprinkler.1/sprinkler.h [hunk 25]
```diff
@@ -19,12 +19,14 @@ extern "C"
 #define SPRINKLER_1_PACKAGE_NAME "sprinkler"
 #define SPRINKLER_1_PACKAGE_VERSION 1
 
-  az_result sprinkler_v1i1_water_now(int32_t area, int32_t timer);
+  az_result sprinkler_v1i1_water_now(int32_t zone, int32_t timer);
```
Comment (smcd253) [inline STMicroelectronics/STM32L4_L4+/samples/sprinkler.1/sprinkler.h:22]: >v1i1 [](http://example.com/codeflow?start=22&length=4)

would this naming scheme need to change with the updated package and interface version formats? this looks like it's following the old standard #Pending

File: STMicroelectronics/STM32L4_L4+/samples/sprinkler.1/sprinkler_interface.c [hunk 26]
```diff
@@ -107,17 +106,6 @@ static az_result sprinkler_1_stop_span_wrapper(az_span model_in_span, az_span* m
   return AZ_ULIB_TRY_RESULT;
 }
```
Comment (smcd253) [inline STMicroelectronics/STM32L4_L4+/samples/sprinkler.1/sprinkler_interface.c:107]: `static az_result sprinkler_v1i1_end(void)` was used as a way to stop the sample on uninstall. is this functionality no longer necessary? #ByDesign

**Review by smcd253 — CHANGES_REQUESTED** [review]
:clock1:

File: DCF_Demo.md (hunk 21 above)
Comment (mamokarz) [reply DCF_Demo.md:360]: No, I'll cover that in the PR that change the capability name to positional, in this case the default will be clearly marked by the `*` in the name.

File: STMicroelectronics/STM32L4_L4+/samples/sprinkler.1/sprinkler.h (hunk 25 above)
Comment (mamokarz) [reply STMicroelectronics/STM32L4_L4+/samples/sprinkler.1/sprinkler.h:22]: Yes. I'll address it as soon as we define how will be the isolation between the autogenerated and developer's code. I created a Feature in DevOpts to track it.

File: STMicroelectronics/STM32L4_L4+/samples/sprinkler.1/sprinkler_interface.c (hunk 26 above)
Comment (mamokarz) [reply STMicroelectronics/STM32L4_L4+/samples/sprinkler.1/sprinkler_interface.c:107]: It is necessary, but it is part of the business logic, not the interface, so I moved it to sprinkler.c.

File: DCF_Demo.md [hunk 27]
```diff
@@ -144,7 +190,7 @@ az iot hub invoke-device-method -n [name-of-iothub] -d [name-of-device] --mn "ip
 
 Turn on the sprinkler, which will be modeled by turning on a LED on the STM Board
 ```
-az iot hub invoke-device-method -n [name-of-iothub] -d [name-of-device] --mn "sprinkler.1.water_now" --mp "{\`"area\`":0}"
+az iot hub invoke-device-method -n [name-of-iothub] -d [name-of-device] --mn "sprinkler.sprinkler.1.water_now" --mp "{\`"zone\`":0}"
```
Comment (mamokarz) [inline DCF_Demo.md]: >{\`"zone\`":0} [](http://example.com/codeflow?start=117&length=14)

Fix command prompt double quotes. #Closed

File: DCF_Demo.md [hunk 28]
```diff
@@ -155,7 +201,7 @@ az iot hub invoke-device-method -n [name-of-iothub] -d [name-of-device] --mn "sp
 
 Turn off the sprinkler, which will be modeled by turning off a LED on the STM Board
 ```
-az iot hub invoke-device-method -n [name-of-iothub] -d [name-of-device] --mn "sprinkler.1.stop" --mp "{\`"area\`":0}" 
+az iot hub invoke-device-method -n [name-of-iothub] -d [name-of-device] --mn "sprinkler.sprinkler.1.stop" --mp "{\`"zone\`":0}"
```
Comment (mamokarz) [inline DCF_Demo.md]: >{\`"zone\`" [](http://example.com/codeflow?start=112&length=11)

Fix command prompt double quotes. #Closed

File: DCF_Demo.md [hunk 29]
```diff
@@ -164,9 +210,9 @@ az iot hub invoke-device-method -n [name-of-iothub] -d [name-of-device] --mn "sp
 }
 ```
 
-We are now sending a message to the device and using the newly installed key_vault to encrypt the message "Welcome to Azure IoT!". The response will be the encrypted result of the message.
+We are now sending a secret message to the device and using the newly installed key_vault to decrypt the message. The response will be the decrypted result of the message.
```
Comment (mamokarz) [inline DCF_Demo.md]: Revert this change. #Closed

File: STMicroelectronics/STM32L4_L4+/samples/key_vault.2/startup/gnu/gcc_setup.s [hunk 30]
```diff
@@ -0,0 +1,162 @@
...
+    ldr     r0, =__reldyn_load_start__
+    sub     r0,r0,r3
+    add     r0,r0,r5
+    ldr     r1, =__reldyn_end__
+    sub     r1,r1,r3
+    add     r1,r1,r5
+relocation_loop:
+    cmp     r0, r1                    // Loop over the rel.dyn table.
+    beq     relocation_done
+    ldr     r2, [r0, #4]              // Table is composed by [offiset, info].
```
Comment (cindydeng1998) [inline STMicroelectronics/STM32L4_L4+/samples/key_vault.2/startup/gnu/gcc_setup.s]: >offiset [](http://example.com/codeflow?start=63&length=7)

offset #Closed

**Comment by cindydeng1998** [comment]
>   while (1)

If this is the default behavior, can we include it in our code so the user doesn't have to include this file everytime?

---
In reply to: [902332045](https://github.com/mamokarz/getting-started/pull/20#issuecomment-902332045) [](http://example.com/codeflow?ancestors=902332045)

---
In reply to: [902332045](https://github.com/mamokarz/getting-started/pull/20#issuecomment-902332045) [](http://example.com/codeflow?ancestors=902332045)

---
Refers to: STMicroelectronics/STM32L4_L4+/samples/sprinkler_v1i1/startup/gnu/precondition_fail.c:10 in a88d44d. [](commit_id = a88d44df08cb7de845da047b7962d8125f0ef7e0, deletion_comment = False)

---
Packages are independent and shall contains its own code. The right way to share a code between packages is using Libraries, for example, what we will do with az_core in the future.

Precondition is an interesting case, it should not exist and we remove it in release mode, however we need it in the package for debug mode. 

Let's keep it as is for now. It is in the "startup" directory that developers shall copy and use as is, they will probably never open this directory to see what is inside.

**Comment by cindydeng1998** [comment]
>   } cipher_1_decrypt_model_in;

typedef az_span cipher_1_decrypt_model_in? since az_ span is already a struct

---
In reply to: [902333013](https://github.com/mamokarz/getting-started/pull/20#issuecomment-902333013) [](http://example.com/codeflow?ancestors=902333013)

---
In reply to: [902333013](https://github.com/mamokarz/getting-started/pull/20#issuecomment-902333013) [](http://example.com/codeflow?ancestors=902333013)

---
Refers to: STMicroelectronics/STM32L4_L4+/samples/key_vault_v2i1/cipher_1_model.h:59 in a88d44d. [](commit_id = a88d44df08cb7de845da047b7962d8125f0ef7e0, deletion_comment = False)

File: STMicroelectronics/STM32L4_L4+/samples/key_vault.1/startup/gnu/gcc_setup.s [hunk 31]
```diff
@@ -0,0 +1,162 @@
...
+    ldr     r0, =__reldyn_load_start__
+    sub     r0,r0,r3
+    add     r0,r0,r5
+    ldr     r1, =__reldyn_end__
+    sub     r1,r1,r3
+    add     r1,r1,r5
+relocation_loop:
+    cmp     r0, r1                    // Loop over the rel.dyn table.
+    beq     relocation_done
+    ldr     r2, [r0, #4]              // Table is composed by [offiset, info].
```
Comment (cindydeng1998) [inline STMicroelectronics/STM32L4_L4+/samples/key_vault.1/startup/gnu/gcc_setup.s]: offset #Closed

**Comment by mamokarz** [comment]
>   while (1)

Packages are independent and shall contains its own code. The right way to share a code between packages is using Libraries, for example, what we will do with az_core in the future.

Precondition is an interesting case, it should not exist and we remove it in release mode, however we need it in the package for debug mode. 

Let's keep it as is for now. It is in the "startup" directory that developers shall copy and use as is, they will probably never open this directory to see what is inside. 

---
In reply to: [902332045](https://github.com/mamokarz/getting-started/pull/20#issuecomment-902332045) [](http://example.com/codeflow?ancestors=902332045,902332045)

---
Refers to: STMicroelectronics/STM32L4_L4+/samples/sprinkler_v1i1/startup/gnu/precondition_fail.c:10 in a88d44d. [](commit_id = a88d44df08cb7de845da047b7962d8125f0ef7e0, deletion_comment = False)

**Comment by mamokarz** [comment]
>   } cipher_1_decrypt_model_in;

Super happy see this comment. :D

---
In reply to: [902333013](https://github.com/mamokarz/getting-started/pull/20#issuecomment-902333013) [](http://example.com/codeflow?ancestors=902333013,902333013)

---
Refers to: STMicroelectronics/STM32L4_L4+/samples/key_vault_v2i1/cipher_1_model.h:59 in a88d44d. [](commit_id = a88d44df08cb7de845da047b7962d8125f0ef7e0, deletion_comment = False)

**Review by smcd253 — APPROVED** [review]
lgtm

**Review by cindydeng1998 — APPROVED** [review]
🌞
